### DISCUSSION TITLE
unable to get vcpkg actually build something in linux using vs-code

### DISCUSSION THREAD

**Discussion by lhksoft** [microsoft/vcpkg#22852] [Q&A]
I give up.
I've got vcpkg working perfectly in windows10 using vstudio 2022 community.
Now I want to port my tryouts in the windows environment into linux.
I even can't anything working right. 

I try to code someting using wxwidgets 3.1.5, installed vcpkg with wxwidgets target x64-linux (gtk). IDE is MS-Code

in vs-code, file <compile_commands.json>
[
{
  "directory": "/home/laurens/Projects/wxTestApp/build",
  "command": "/bin/x86_64-linux-gnu-g++-9  -DWXUSINGDLL -D__WXGTK3__ -D__WXGTK__ -isystem 
/opt/vcpkg/packages/wxwidgets_x64-linux/lib/wx/include/gtk3-unicode-static-3.1 -isystem /opt/vcpkg/packages/wxwidgets_x64-linux/include/wx-3.1 -isystem /opt/vcpkg/installed/x64-linux/include   -pthread -g   -pthread -o CMakeFiles/wxTestApp.dir/src/main.cpp.o -c /home/laurens/Projects/wxTestApp/src/main.cpp",
  "file": "/home/laurens/Projects/wxTestApp/src/main.cpp"
}
]

main CMakeLists.txt file in root of some project
cmake_minimum_required(VERSION 3.0.0)
project(wxTestApp VERSION 0.1.0)
include(/opt/vcpkg/scripts/buildsystems/vcpkg.cmake)

set(
  PROG_SRCS
  src/main.cpp
)

find_package(wxWidgets REQUIRED COMPONENTS net core base)
include(${wxWidgets_USE_FILE})

add_executable(${PROJECT_NAME} ${PROG_SRCS})

target_link_libraries(${PROJECT_NAME} ${wxWidgets_LIBRARIES})

Then when compiling, build always fails with library's it cannot find (but I checked, the library paths are given)

[main] Building folder: wxTestApp 
[build] Starting build
[proc] Executing command: /usr/bin/cmake --build /home/laurens/Projects/wxTestApp/build --config Debug --target wxTestApp -j 6 --
[build] /usr/bin/cmake -S/home/laurens/Projects/wxTestApp -B/home/laurens/Projects/wxTestApp/build --check-build-system CMakeFiles/Makefile.cmake 0
[build] Re-run cmake file: Makefile older than: /opt/vcpkg/scripts/buildsystems/vcpkg.cmake
[build] -- Configuring done
[build] -- Generating done
[build] -- Build files have been written to: /home/laurens/Projects/wxTestApp/build
[build] /usr/bin/make -f CMakeFiles/Makefile2 wxTestApp
[build] make[1]: Entering directory '/home/laurens/Projects/wxTestApp/build'
[build] /usr/bin/cmake -S/home/laurens/Projects/wxTestApp -B/home/laurens/Projects/wxTestApp/build --check-build-system CMakeFiles/Makefile.cmake 0
[build] /usr/bin/cmake -E cmake_progress_start /home/laurens/Projects/wxTestApp/build/CMakeFiles 2
[build] /usr/bin/make -f CMakeFiles/Makefile2 CMakeFiles/wxTestApp.dir/all
[build] make[2]: Entering directory '/home/laurens/Projects/wxTestApp/build'
[build] /usr/bin/make -f CMakeFiles/wxTestApp.dir/build.make CMakeFiles/wxTestApp.dir/depend
[build] make[3]: Entering directory '/home/laurens/Projects/wxTestApp/build'
[build] cd /home/laurens/Projects/wxTestApp/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/laurens/Projects/wxTestApp /home/laurens/Projects/wxTestApp /home/laurens/Projects/wxTestApp/build /home/laurens/Projects/wxTestApp/build /home/laurens/Projects/wxTestApp/build/CMakeFiles/wxTestApp.dir/DependInfo.cmake --color=
[build] make[3]: Leaving directory '/home/laurens/Projects/wxTestApp/build'
[build] /usr/bin/make -f CMakeFiles/wxTestApp.dir/build.make CMakeFiles/wxTestApp.dir/build
[build] make[3]: Entering directory '/home/laurens/Projects/wxTestApp/build'
[build] [ 50%] Linking CXX executable wxTestApp
[build] /usr/bin/cmake -E cmake_link_script CMakeFiles/wxTestApp.dir/link.txt --verbose=1
[build] /bin/x86_64-linux-gnu-g++-9   -pthread -g  -rdynamic CMakeFiles/wxTestApp.dir/src/main.cpp.o  -o wxTestApp   -L/opt/vcpkg/packages/wxwidgets_x64-linux/lib  -Wl,-rpath,/opt/vcpkg/packages/wxwidgets_x64-linux/lib -L/opt/vcpkg/packages/wxwidgets_x64-linux/lib -pthread /opt/vcpkg/packages/wxwidgets_x64-linux/lib/libwx_baseu_net-3.1.a /opt/vcpkg/packages/wxwidgets_x64-linux/lib/libwx_gtk3u_core-3.1.a /opt/vcpkg/packages/wxwidgets_x64-linux/lib/libwx_baseu-3.1.a -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lSM -lICE -lX11 -lXext -lXtst -lwx_baseu-3.1 -llibjpeg.a> -llibjpeg.a> -llibpng.a> -llibpng16d.a> -llibz.a> -llibz.a> -lm -llibtiff.a> -llibtiffd.a> -lliblzma.a> -lliblzmad.a> -llibjpeg.a> -llibjpeg.a> -llibz.a> -llibz.a> -lm -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lSM -lICE -lX11 -lXext -lXtst -llibz.a> -llibz.a> -lwxregexu-3.1 -lc -pthread -ldl /opt/vcpkg/installed/x64-linux/debug/lib/libtiffd.a /opt/vcpkg/installed/x64-linux/debug/lib/libexpat.a /opt/vcpkg/installed/x64-linux/debug/lib/libz.a /opt/vcpkg/installed/x64-linux/debug/lib/libpng16d.a -lwx_baseu-3.1 -llibpng.a> -llibpng16d.a> -llibtiff.a> -llibtiffd.a> -lliblzma.a> -lliblzmad.a> -lwxregexu-3.1 -lc -ldl /opt/vcpkg/installed/x64-linux/debug/lib/liblzmad.a -pthread /opt/vcpkg/installed/x64-linux/debug/lib/libjpeg.a -lm /opt/vcpkg/installed/x64-linux/debug/lib/libz.a 
[build] /usr/bin/ld: cannot find -llibjpeg.a>
[build] /usr/bin/ld: cannot find -llibjpeg.a>
[build] /usr/bin/ld: cannot find -llibpng.a>
[build] /usr/bin/ld: cannot find -llibpng16d.a>
[build] /usr/bin/ld: cannot find -llibz.a>
[build] /usr/bin/ld: cannot find -llibz.a>
[build] /usr/bin/ld: cannot find -llibtiff.a>
[build] /usr/bin/ld: cannot find -llibtiffd.a>
[build] /usr/bin/ld: cannot find -lliblzma.a>
[build] /usr/bin/ld: cannot find -lliblzmad.a>
[build] /usr/bin/ld: cannot find -llibjpeg.a>
[build] /usr/bin/ld: cannot find -llibjpeg.a>
[build] /usr/bin/ld: cannot find -llibz.a>
[build] /usr/bin/ld: cannot find -llibz.a>
[build] collect2: error: ld returned 1 exit status
[build] make[3]: *** [CMakeFiles/wxTestApp.dir/build.make:97: wxTestApp] Error 1
[build] make[2]: *** [CMakeFiles/Makefile2:79: CMakeFiles/wxTestApp.dir/all] Error 2
[build] make[1]: *** [CMakeFiles/Makefile2:86: CMakeFiles/wxTestApp.dir/rule] Error 2
[build] make: *** [Makefile:143: wxTestApp] Error 2
[build] make[3]: Leaving directory '/home/laurens/Projects/wxTestApp/build'
[build] make[2]: Leaving directory '/home/laurens/Projects/wxTestApp/build'
[build] make[1]: Leaving directory '/home/laurens/Projects/wxTestApp/build'
[build] Build finished with exit code 2

It are those '[build] /usr/bin/ld: cannot find -lXXX.a' I cannot resolve.

I search the Internet, found many solutions but nothing of them works for me.
After days searching and trying, I give up. Unless someone can help me with this, getting vcpkg in ms-code in linux-x64 using wxwidgets to actually work. VCPKG is okay when using Studio in Win10, but appearently it's a big pain to get it to work in linux (-Mint) using ms-code. Pitty

**Comment by dg0yt** [comment]
a) vcpkg's port wxwidgets has a number of issues in Linux. Try something simpler first.

b) 
~~~
project(wxTestApp VERSION 0.1.0)
include(/opt/vcpkg/scripts/buildsystems/vcpkg.cmake)
~~~

`vcpkg.cmake` must be used as cmake toolchain file. It must be set on the first invocation of cmake, before the first call to the project command.

**Comment by lhksoft** [comment]
I did using the cmake-toolchain before but that didn't do much either.

Well, I've done a workaround.
I downloaded gtk3 + gtk3-dev and installed those inside my linux-mint (while I'm not sure whether Mint still uses GTK2 or already GTK3). Then I downloaded the sources for wxWidgets3.1.5 and compiled it manually. Then installed it into my Mint using `make install`. 
So completely without vcpkg in linux. I code at the same project in win10 using VStudio and in there I use vcpkg, but in linux it's only with ms-code and my own compiled and installed wxwidgets without vcpkg.

Now my code is cross-platform and it works like I want it to be.
I think vcpkg still has too much problems to use in linux with ms-code.

**Reply by dg0yt** [reply]
> I did using the cmake-toolchain before but that didn't do much either.

I explicitly tried to mark this as a separate issue. But it is an issue.

> I think vcpkg still has too much problems to use in linux with ms-code.

Not sure what you meant with "linux in ms-code".

For vcpkg port wxwidgets: 
WxWidgets still requires GTK3 on Linux. However, GTK was upgraded in vcpkg from 3 to 4 without taking care of WxWidgets. Using system GTK3 is probably error-prone due to systems pulling in different sets of other libraries via dynamic linkage. (There is a PR to add gtk3 again.) 
In addition, there are issues with properly separating debug from release configuration. This is a native problem in vcpkg, but unusual for Linux developers.

**Reply by lhksoft** [reply]
> > I think vcpkg still has too much problems to use in linux with ms-code.
> 
> Not sure what you meant with "linux in ms-code".

Well, it is like I say, but maybe my English is not so swell, so sorry for that.
What I ment, I code in MS-Code (while I'm used to Visual Studio, MS-Code looks most like Studio), and I do it in Linux-Mint

> For vcpkg port wxwidgets:

I thank you for your answer.
Now I know it's not my doing that raise the problems, but truly from within vcpkg which upgraded to GTK4 and currently wxwidgets only works with GTK3+ ..

I'll keep coding without vcpkg (in linux) and with it in win10.

Hopefully, someday this will be fixed?
